[SPARK-14163][CORE] SumEvaluator and countApprox cannot reliably handle RDDs of size 1.

yongtang opened the pull request:
## What changes were proposed in this pull request?

This fix fixes issues in SPARK-14163 where SumEvaluator could not handle `counter.count <=` 1 as `degreesOfFreedom` requires `counter.count > 1`.
In this fix, `counter.count <= 1` is handled separately.
## How was this patch tested?

A manual test was done to make sure that no Exception is thrown for `degreesOfFreedom`.

yongtang committed:
[SPARK-14163][CORE] SumEvaluator and countApprox cannot reliably handle RDDs of size 1.

This fix fixes issues in SPARK-14163 where SumEvaluator could not handle
counter.count of `<=` 1 as degreesOfFreedom requires counter.count > 1.

AmplabJenkins commented:
Can one of the admins verify this patch?

marcintustin commented on core/src/main/scala/org/apache/spark/partial/SumEvaluator.scala:
```diff
@@ -42,6 +42,14 @@ private[spark] class SumEvaluator(totalOutputs: Int, confidence: Double)
     } else if (outputsMerged == 0) {
       new BoundedDouble(0, 0.0, Double.NegativeInfinity, Double.PositiveInfinity)
+    } else if (counter.count == 0) {
+      new BoundedDouble(0, 0.0, Double.NegativeInfinity, Double.PositiveInfinity)
```
Why not just use an || to avoid repeating code?

yongtang replied:
Thanks. Just updated the pull request.

srowen commented on core/src/main/scala/org/apache/spark/partial/SumEvaluator.scala:
```diff
@@ -42,6 +42,14 @@ private[spark] class SumEvaluator(totalOutputs: Int, confidence: Double)
+    } else if (counter.count == 0) {
+      new BoundedDouble(0, 0.0, Double.NegativeInfinity, Double.PositiveInfinity)
+    } else if (counter.count == 1) {
+      val p = outputsMerged.toDouble / totalOutputs
```
This duplicates code in the following block. I think you can compute the common values in one shared block and then branch afterwards to handle count == 1.

yongtang replied:
Thanks @srowen just updated the pull request.

marcintustin commented:
FYI I have a more parsimonious change here: #11982

yongtang committed:
[SPARK-14163][CORE] SumEvaluator and countApprox cannot reliably handle RDDs of size 1 (continue).

This fix merge duplicate code in the block (`outputsMerged == 0 || counter.count == 0`).

srowen commented:
@yongtang do you mind closing this PR in favor of https://github.com/apache/spark/pull/12016 ?

yongtang commented:
@srowen Sure let me close this PR.
